doc/wg/core/notes/core-notes-2021-07-09.md

```diff
@@ -0,0 +1,157 @@
+* Brad: So we could have some written classification. That would also help us keep track of this and eventually to phase out boards that aren't being tested anymore.
+* Hudson: Last comment on this is that we don't want to block 2.0 on stuff that hasn't been submitted or won't be all that quick if it's not ABI blocking. But, maybe we should look at the list of outstanding PRs and look for things to force through before release. My opinion is that we shouldn't do that this time, but maybe others disagree.
+* Branden: I agree with that strategy. 2.1 can be all the great updates to Tock, while 2.0 is the ABI changes and soundness issues.
+* Brad: I didn't see anything in particular I want to include.
+* Hudson: We can still merge things that are ready, of course.
+
+## Appslice swapping in capsules
+* Brad: Related question, I'm still confused about what we're promising to userspace around appslices and swapping. TRD says if you call allow you'll get the old buffer back. We're not enforcing that and leaving it as a TODO, which I think makes sense. But are we violating the TRD104 promises? Are we going to handle it soon? I think we need a record of where we stand.
+* Leon: I think this is a valid point and it's confusing because of how it developed. Hudson and I planned on enforcing restrictions before Tock 2.0, which is why it's written in the TRD104 this way. However, we've since decided we don't have to enforce for soundness and it's okay for now. But it would still be a good TODO. The TRD currently has expectations we don't meet in our codebase, so we should edit the TRD to say our expectation and our implementation.
+* Hudson: I thought we had said capsules were allowed to refuse to return a buffer. For example if we were passing it into a DMA request
+* Leon: Well, we can't use userspace buffers for DMA anyways. Capsules can refuse operations, for example if there is a complex piece of hardware. So if the buffer only goes away if the process dies, that makes things easier to handle. Otherwise _every_ operation would have to be abortable.
+* Hudson: I thought there was also the idea that a capsule can store info in an allowed buffer and could rely on it still being there. So we didn't want to let processes revoke buffers at any time.
+* Leon: Yes, but it can't be trusted, since the process could change data in the buffer whenever it wants.
+* Hudson: The reason I'm asking is that maybe the text in TRD104 is incorrect if we permit capsules to refuse an allow call. Which I thought was originally the plan.
+* Leon: I'm confused.
+* Hudson: Currently the documentation of TRD104, I believe, is that if userspace requests a buffer back, it must get it back. But that doesn't sync with my memory that allows can be refused.
+* Leon: I think the wording is that _if_ it gets a buffer back, it must be the right one. And the "must" restriction there is what we don't enforce.
```

Comment:
Yes. This is stating a requirement for the implementation of a capsule. That's why it's a MUST. If the document just said that the kernel always does it, then we could just make that true. Put another way, a capsule that doesn't follow this MUST is not compliant with 104.

Reply:
Agreed. I suppose there are two perspectives, which are a great source of confusion when they're mixed:
- the capsule implementation: a capsule MUST return the correct buffer to be compliant
- the system call / kernel behavior: applications MUST NOT rely on previously allowed buffers to be returned, as least w.r.t. to the current kernel implementation.

Unfortunately, the TRD104 currently states: _Calling a Read-Write Allow system call returns a buffer (address and length).  On the first call to a Read-Write Allow system call, the kernel returns a zero-length buffer. Subsequent successful calls to Read-Write Allow return the previous buffer passed._

This, to me, seems very much written from an ABI perspective, outlining the behavior of the kernel in response to an application's actions. And here, we definitely do not want to give the guarantee that the previous buffers is going to be returned in every case, because we're currently not enforcing that.

I will open a PR to rephrase this so that it's clear that (a) a capsule must return the previously passed in buffer, and (b) that an application cannot simply rely on that fact, in the sense of it being enforced by the kernel.

Reply:
Just opened #2692 which tried to incorporate these details into the specification. Sorry it took me a bit longer.